src/build/esbuild.ts

```diff
@@ -0,0 +1,186 @@
+import {
+  denoPlugins,
+  esbuild,
+  esbuildTypes,
+  esbuildWasmURL,
+  fromFileUrl,
+  toFileUrl,
+} from "./deps.ts";
+import { Builder, BuildSnapshot } from "./mod.ts";
+
+export interface EsbuildBuilderOptions {
+  /** The build ID. */
+  buildID: string;
+  /** The entrypoints, mapped from name to URL. */
+  entrypoints: Record<string, string>;
+  /** Whether or not this is a dev build. */
+  dev: boolean;
+  /** The import map URL. */
+  importMapURL: URL;
+  /** The JSX configuration. */
+  jsxConfig: JSXConfig;
+}
+
+export interface JSXConfig {
+  jsx: "react" | "react-jsx";
+  jsxImportSource?: string;
+}
+
+export class EsbuildBuilder implements Builder {
+  #options: EsbuildBuilderOptions;
+
+  constructor(options: EsbuildBuilderOptions) {
+    this.#options = options;
+  }
+
+  async build(): Promise<EsbuildSnapshot> {
+    const opts = this.#options;
+    try {
+      await initEsbuild();
+
+      const absWorkingDir = Deno.cwd();
+
+      // In dev-mode we skip identifier minification to be able to show proper
+      // component names in Preact DevTools instead of single characters.
+      const minifyOptions: Partial<esbuildTypes.BuildOptions> = opts.dev
+        ? {
+          minifyIdentifiers: false,
+          minifySyntax: true,
+          minifyWhitespace: true,
+        }
+        : { minify: true };
+
+      const bundle = await esbuild.build({
+        entryPoints: opts.entrypoints,
+
+        platform: "neutral",
+        target: ["chrome99", "firefox99", "safari15"],
+
+        format: "esm",
+        bundle: true,
+        splitting: true,
+        treeShaking: true,
+        sourcemap: opts.dev ? "linked" : false,
+        ...minifyOptions,
+
+        jsx: JSX_RUNTIME_MODE[opts.jsxConfig.jsx],
+        jsxImportSource: opts.jsxConfig.jsxImportSource,
+
+        absWorkingDir,
+        outdir: ".",
+        write: false,
+        metafile: true,
+
+        plugins: [
+          buildIdPlugin(opts.buildID),
+          ...denoPlugins({ importMapURL: opts.importMapURL.href }),
+        ],
+      });
+
+      const files = new Map<string, Uint8Array>();
+      const dependencies = new Map<string, string[]>();
+
+      const absWorkingDirLen = toFileUrl(absWorkingDir).href.length + 1;
+
+      for (const file of bundle.outputFiles) {
+        const path = toFileUrl(file.path).href.slice(absWorkingDirLen);
+        files.set(path, file.contents);
+      }
+
+      const metaOutputs = new Map(Object.entries(bundle.metafile.outputs));
+
+      // deno-lint-ignore no-inner-declarations
+      function walk(path: string, seen: Set<string>) {
+        if (seen.has(path)) return;
+        seen.add(path);
```

Comment:
Is recursion necessary here? From what I've seen, esbuild constructs the dependency tree in such a way that no recursion is required.

Reply:
It is not guaranteed that the deps are always entrypoint -> chunk. I think entrypoint -> chunk -> chunk is possible

Reply:
It is not guaranteed that the deps are always entrypoint -> chunk. I think entrypoint -> chunk -> chunk is possible

Reply:
It is not guaranteed that the deps are always entrypoint -> chunk. I think entrypoint -> chunk -> chunk is possible

Reply:
It is not guaranteed that the deps are always entrypoint -> chunk. I think entrypoint -> chunk -> chunk is possible

Reply:
It looks like the entry point contains the data for all the chunks it depends on so that no chunk->chunk reads occur.

(For example, `chunk-E7WS2GRN.js` is not used in the entry point like below, but it is imported here because it is imported from `chunk-OYKCTXOR.js`.)
![image](https://github.com/denoland/fresh/assets/40050810/6adec24c-bbb9-4b02-ba58-cc877719c293)

That being said, these things aren't documented in esbuild, so I could be wrong!

Reply:
I see - I think you're right. We can remove the logic then